adapters/akcelo/akcelo.go

```
@@ -0,0 +1,191 @@
+	"github.com/prebid/prebid-server/v3/config"
+	"github.com/prebid/prebid-server/v3/errortypes"
+	"github.com/prebid/prebid-server/v3/openrtb_ext"
+	"github.com/prebid/prebid-server/v3/util/jsonutil"
+	"github.com/prebid/prebid-server/v3/util/ptrutil"
+	"net/http"
+	"net/url"
+)
+
+type adapter struct {
+	uri *url.URL
+}
+
+type extObj struct {
+	Akcelo openrtb_ext.ExtImpAkcelo `json:"akcelo"`
+}
+
+var NoValidSiteIdError = &errortypes.BadInput{Message: "Cannot find valid siteId"}
+var NoValidImpError = &errortypes.BadInput{Message: "No valid Imp"}
+
+func Builder(bidderName openrtb_ext.BidderName, config config.Adapter, server config.Server) (adapters.Bidder, error) {
+	uri, err := url.Parse(config.Endpoint)
+	if err != nil {
+		return nil, err
+	}
+	bidder := &adapter{uri: uri}
+	return bidder, nil
+}
+
+func (a *adapter) MakeRequests(request *openrtb2.BidRequest, reqInfo *adapters.ExtraRequestInfo) ([]*adapters.RequestData, []error) {
+	requestData, errs := a.prepareBidRequest(request)
+	return requestData, errs
+}
+
+func (a *adapter) MakeBids(internalRequest *openrtb2.BidRequest, externalRequest *adapters.RequestData, response *adapters.ResponseData) (*adapters.BidderResponse, []error) {
+	if adapters.IsResponseStatusCodeNoContent(response) {
+		return nil, nil
+	}
+	if err := adapters.CheckResponseStatusCodeForErrors(response); err != nil {
+		return nil, []error{err}
+	}
+	var bidResponse openrtb2.BidResponse
+	if err := jsonutil.Unmarshal(response.Body, &bidResponse); err != nil {
+		return nil, []error{err}
+	}
+	return extractBids(bidResponse)
+}
+
+func (a *adapter) prepareBidRequest(bidRequest *openrtb2.BidRequest) ([]*adapters.RequestData, []error) {
+	bidRequest = createSitePublisher(bidRequest)
+	requestData := make([]*adapters.RequestData, 0, 1)
+
+	headers := http.Header{}
+	headers.Add("Content-Type", "application/json;charset=utf-8")
+	headers.Add("Accept", "application/json")
+
+	if len(bidRequest.Imp) == 0 {
```

In order to fail fast and to avoid that the process that happens inside `createSitePublisher(bidRequest)` just to goes to waste when `len(bidRequest.Imp) == 0`, can we please move this check to the beginning of `prepareBidRequest()`? 
```
56   func (a *adapter) prepareBidRequest(bidRequest *openrtb2.BidRequest) ([]*adapters.RequestData, []error) {
   +     if len(bidRequest.Imp) == 0 {
   +         return nil, []error{NoValidImpError}
   +     }
57       bidRequest = createSitePublisher(bidRequest)
58       requestData := make([]*adapters.RequestData, 0, 1)
59  
60       headers := http.Header{}
61       headers.Add("Content-Type", "application/json;charset=utf-8")
62       headers.Add("Accept", "application/json")
63  
64 -     if len(bidRequest.Imp) == 0 {
65 -         return nil, []error{NoValidImpError}
66 -     }
67  
68       var errs []error
69       for i := range bidRequest.Imp {
70           imp := &bidRequest.Imp[i]
adapters/akcelo/akcelo.go
```

---

Fixed